nominal/core/data_review.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from datetime import datetime, timedelta
+from time import sleep
+from typing import Protocol, Sequence
+
+from typing_extensions import Self
+
+from nominal._api.scout_service_api import scout_checks_api, scout_datareview_api, scout_integrations_api
+from nominal.core import checklist
+from nominal.core._clientsbunch import HasAuthHeader
+from nominal.core._utils import HasRid
+from nominal.ts import IntegralNanosecondsUTC, _SecondsNanos
+
+
+@dataclass(frozen=True)
+class DataReview(HasRid):
+    rid: str
+    run_rid: str
+    checklist_rid: str
+    checklist_commit: str
+    completed: bool
+
+    _clients: _Clients = field(repr=False)
+
+    class _Clients(HasAuthHeader, Protocol):
+        @property
+        def datareview(self) -> scout_datareview_api.DataReviewService: ...
+
+    @classmethod
+    def _from_conjure(cls, clients: _Clients, data_review: scout_datareview_api.DataReview) -> Self:
+        executing_states = [
+            check.automatic_check.state._pending_execution or check.automatic_check.state._executing
+            for check in data_review.checklist.checks
+            if check.automatic_check
+        ]
+        completed = not any(executing_states)
+        return cls(
+            rid=data_review.rid,
+            run_rid=data_review.run_rid,
+            checklist_rid=data_review.checklist.checklist.rid,
+            checklist_commit=data_review.checklist.checklist.commit,
+            completed=completed,
+            _clients=clients,
+        )
+
+    def get_violations(self) -> Sequence[CheckViolation]:
+        """Retrieves the list of check violations for the data review."""
+        response = self._clients.datareview.get_check_alerts_for_data_review(self._clients.auth_header, self.rid)
+        return [CheckViolation._from_conjure(alert) for alert in response]
+
+
+@dataclass(frozen=True)
+class CheckViolation:
+    rid: str
+    check_rid: str
+    name: str
+    start: IntegralNanosecondsUTC
+    end: IntegralNanosecondsUTC | None
+    priority: checklist.Priority | None
+
+    @classmethod
+    def _from_conjure(cls, check_alert: scout_datareview_api.CheckAlert) -> CheckViolation:
+        return cls(
+            rid=check_alert.rid,
+            check_rid=check_alert.check_rid,
+            name=check_alert.name,
+            start=_SecondsNanos.from_api(check_alert.start).to_nanoseconds(),
+            end=_SecondsNanos.from_api(check_alert.end).to_nanoseconds() if check_alert.end is not None else None,
+            priority=checklist._conjure_priority_to_priority(check_alert.priority)
+            if check_alert.priority is not scout_checks_api.Priority.UNKNOWN
+            else None,
+        )
+
+
+@dataclass(frozen=True)
+class DataReviewBatchBuilder:
+    integration_rids: list[str]
+    _requests: list[scout_datareview_api.CreateDataReviewRequest]
```

Comment:
why is one of these public and the other internal? since users aren't really meant to edit, lets mark them both as internal I guess?

Reply:
if we want them to be public, we should not use `list` but a more generic sequence type